Libraries/LibWeb/Parser/HTMLDocumentParser.cpp

```diff
@@ -116,18 +116,39 @@ void HTMLDocumentParser::process_using_the_rules_for(InsertionMode mode, HTMLTok
 
 void HTMLDocumentParser::handle_initial(HTMLToken& token)
 {
-    if (token.type() == HTMLToken::Type::DOCTYPE) {
+    if (token.is_character() && token.is_parser_whitespace()) {
+        return;
+    }
+    
+    if(token.is_comment()) {
+        auto comment = adopt(*new Comment(document(), token.m_comment_or_character.data.to_string()));
+        document().append_child(move(comment));
+        return;
+    }
+
+    if (token.is_doctype()) {
         auto doctype = adopt(*new DocumentType(document()));
         doctype->set_name(token.m_doctype.name.to_string());
         document().append_child(move(doctype));
         m_insertion_mode = InsertionMode::BeforeHTML;
         return;
     }
+
     ASSERT_NOT_REACHED();
 }
 
 void HTMLDocumentParser::handle_before_html(HTMLToken& token)
 {
+    if(token.is_doctype()) {
+        PARSE_ERROR();
```

Comment:
Note that `PARSE_ERROR()` does not automatically return, since there are sometimes steps to take after a parse error.
Because of this, we have to `return;` here, otherwise we will fall down into the "anything else" section below :)